Add HIDDevice.forget() method

beaufortfrancois opened the pull request:
This PR is an attempt to add a way for web developers to revoke permission access to a paired HIDDevice. See https://github.com/WICG/webhid/issues/39

```js
// Request a HID device.
const [device] = await navigator.hid.requestDevice({ filters: [] });

// Then later... revoke permission to the HID device.
await device.forget();
```

@nondebug Please have a look

nondebug commented on index.html:
```diff
@@ -1642,6 +1648,43 @@ <h3>
+        <ol>
+          <li>Let |promise:Promise| be [=a new promise=].
+          </li>
+          <li>Set [=this=].{{HIDDevice/[[state]]}} to `"forgetting"`.
```
The device might be open or closed. If it's open, I think we should close it (and issue relevant callbacks) as part of these steps.

beaufortfrancois replied:
By closing it when forgetting it, we're introducing race conditions in the spec as `{{HIDDevice/[[state]]}}` would become `"closing"`, not `"forgetting"` anymore.
What would be the best way to handle this then?

nondebug replied:
I think this is fine now that we invoke pending callbacks on forgetting.

nondebug commented on index.html:
```diff
@@ -1590,6 +1592,10 @@ <h3>
           <li>Let |promise:Promise| be [=a new promise=].
           </li>
+          <li>If [=this=].{{HIDDevice/[[state]]}} is either `"forgotten"` or
+          `"forgetting"`, then [=reject=] |promise| with a
```
Currently close() never rejects and I think it would be nice to keep it that way if we can. If close() is called on a "forgotten" or "forgetting" device let's resolve instead of rejecting so developers don't have to handle it like an error.

beaufortfrancois replied:
`close()` actually rejects if `open()` is not finished yet in Chromium implementation.

```js
device.open();
device.close();
// Promise {<rejected>: DOMException: An operation that changes the device state is in progress.}
```

Resolving `close()` while the device is actually not accessible seems weird to me.

nondebug commented on index.html:
```diff
@@ -1642,6 +1648,43 @@ <h3>
+          </li>
+          <li>Perform the following steps [=in parallel=]:
+            <ol>
+              <li>Invoke the user agent to revoke access to all the [=HID interfaces=]
```
I assume HIDDevice instances for other HID interfaces would also eventually become "forgotten" but currently there's no mechanism for this to happen. Let's do that here by iterating over all of {{HID/[[devices]]}} and repeating the "forgetting" steps for each HIDDevice that shares device access permissions with the original device.

beaufortfrancois replied:
Done. Thanks!

nondebug commented on index.html:
```diff
@@ -962,7 +962,8 @@ <h3>
                   <li>If the user has allowed the site to access |device| as
-                  the result of a previous call to {{HID/requestDevice()}},
+                  the result of a previous call to {{HID/requestDevice()}} and
+                  |device|.{{HIDDevice/[[state]]}} is not `"forgotten"`,
```
{{HID/requestDevice()}} should reset {{HIDDevice/[[state]]}} to "closed" for any returned HIDDevice instances in case the devices were previously forgotten.

beaufortfrancois replied:
You're absolutely right. Good catch!

beaufortfrancois commented:
@nondebug Thanks for reviewing! I've addressed some of your feedback in https://github.com/WICG/webhid/pull/84/commits/32bdda857212e1cc03e39a8cd8498c5625e5d3d7.

beaufortfrancois commented:
FYI I've started a Chromium CL at https://chromium-review.googlesource.com/c/chromium/src/+/3297868.

reillyeon commented on index.html:
```diff
@@ -1642,6 +1649,52 @@ <h3>
+                  </li>
+                  <li>Set |device|.{{HIDDevice/[[state]]}} to `"forgetting"`.
+                  </li>
+                  <li>Invoke the user agent to revoke access to all the [=HID interfaces=]
```
This should reject all pending promises the way that `close()` does.

beaufortfrancois replied:
Done. Thanks!

beaufortfrancois commented:
@nondebug Shall we merge this PR?

beaufortfrancois commented:
@nondebug I'm about to commit https://chromium-review.googlesource.com/c/chromium/src/+/3297868. Can you merge this spec PR?
